FIAT/functional.py

```diff
@@ -543,52 +693,103 @@ def __init__(self, ref_el, v, w, p):
+        weights = Q.get_weights()
+        pt_dict = OrderedDict()
+        for pt, wgt, phi in zip(pts, weights, P_at_qpts):
+            pt_dict[pt] = [(wgt*phi*n[i], (i, )) for i in range(sd)]
+        super().__init__(ref_el, (sd, ), pt_dict, {}, "IntegralMomentOfScaledNormalEvaluation")
 
-        assert len(f_at_qpts.shape) == 2
-        assert f_at_qpts.shape[0] == 2
-        assert f_at_qpts.shape[1] == nqp
 
+class IntegralMomentOfTangentialEvaluation(Functional):
+    r"""
+    \int_F v\cdot n p ds
+    p \in Polynomials
+    :arg ref_el: reference element for which F is a codim-1 entity
+    :arg Q: quadrature rule on the face
+    :arg P_at_qpts: polynomials evaluated at quad points
+    :arg facet: which facet.
+    """
+    def __init__(self, ref_el, Q, P_at_qpts, facet):
+        # scaling on the tangent is ok because edge length then weights
+        # the reference element quadrature appropriately
         sd = ref_el.get_spatial_dimension()
+        assert sd == 2
+        t = ref_el.compute_edge_tangent(facet)
+        transform = ref_el.get_entity_transform(sd - 1, facet)
+        pts = tuple(map(lambda p: tuple(transform(p)), Q.get_points()))
+        weights = Q.get_weights()
+        pt_dict = OrderedDict()
+        for pt, wgt, phi in zip(pts, weights, P_at_qpts):
+            pt_dict[pt] = [(wgt*phi*t[i], (i, )) for i in range(sd)]
+        super().__init__(ref_el, (sd, ), pt_dict, {}, "IntegralMomentOfScaledTangentialEvaluation")
 
-        dpt_dict = OrderedDict()
-
-        alphas = [tuple([1 if j == i else 0 for j in range(sd)]) for i in range(sd)]
-        for q, pt in enumerate(dpts):
-            dpt_dict[tuple(pt)] = [(qwts[q]*f_at_qpts[i, q], alphas[j], (i, j)) for i in range(2) for j in range(2)]
 
-        Functional.__init__(self, ref_el, tuple(),
-                            {}, dpt_dict, "IntegralMomentOfTensorDivergence")
+class IntegralMomentOfNormalNormalEvaluation(Functional):
+    r"""
+    \int_F (n^T tau n) p ds
+    p \in Polynomials
+    :arg ref_el: reference element for which F is a codim-1 entity
+    :arg Q: quadrature rule on the face
+    :arg P_at_qpts: polynomials evaluated at quad points
+    :arg facet: which facet.
+    """
+    def __init__(self, ref_el, Q, P_at_qpts, facet):
+        # scaling on the normal is ok because edge length then weights
+        # the reference element quadrature appropriately
+        n = ref_el.compute_scaled_normal(facet)
+        sd = ref_el.get_spatial_dimension()
+        transform = ref_el.get_entity_transform(sd - 1, facet)
+        pts = tuple(map(lambda p: tuple(transform(p)), Q.get_points()))
+        weights = Q.get_weights()
+        pt_dict = OrderedDict()
+        for pt, wgt, phi in zip(pts, weights, P_at_qpts):
+            pt_dict[pt] = [(wgt*phi*n[i], (i, )) for i in range(sd)]
+        super().__init__(ref_el, (sd, ), pt_dict, {}, "IntegralMomentOfScaledNormalEvaluation")
```

Comment:
A few of these superclass init calls have a functional name that doesn't match the class name. AFAICT this name is just an arbitrary label that one attaches, so it doesn't have to match.

To avoid this, we could make it an optional argument, and do:

```python
class Functional:
    def __init__(self, ..., functional_type=None):
        ...
        self.functional_type = functional_type or type(self).__name__
```

But maybe better done as a separate change.